crates/bevy_ecs/src/world/mod.rs

```diff
@@ -0,0 +1,819 @@
+    /// Resources are "unique" data of a given type.
+    /// # Safety
+    /// This will allow aliased mutable access to the given resource type. The caller must ensure that only
+    /// one mutable access exists at a time.
+    #[inline]
+    pub unsafe fn get_resource_unchecked_mut<T: Component>(&self) -> Option<Mut<'_, T>> {
+        let component_id = self.components.get_resource_id(TypeId::of::<T>())?;
+        self.get_resource_unchecked_mut_with_id(component_id)
+    }
+
+    /// Gets a reference to the non-send resource of the given type, if it exists. Otherwise returns [None]
+    /// Resources are "unique" data of a given type.
+    #[inline]
+    pub fn get_non_send_resource<T: 'static>(&self) -> Option<&T> {
+        let component_id = self.components.get_resource_id(TypeId::of::<T>())?;
+        // SAFE: component id matches type T
+        unsafe { self.get_non_send_with_id(component_id) }
+    }
+
+    /// Gets a mutable reference to the non-send resource of the given type, if it exists. Otherwise returns [None]
+    /// Resources are "unique" data of a given type.
+    #[inline]
+    pub fn get_non_send_resource_mut<T: 'static>(&mut self) -> Option<Mut<'_, T>> {
+        // SAFE: unique world access
+        unsafe { self.get_non_send_resource_unchecked_mut() }
+    }
+
+    /// Gets a mutable reference to the non-send resource of the given type, if it exists. Otherwise returns [None]
+    /// Resources are "unique" data of a given type.
+    /// # Safety
+    /// This will allow aliased mutable access to the given non-send resource type. The caller must ensure that only
+    /// one mutable access exists at a time.
+    #[inline]
+    pub unsafe fn get_non_send_resource_unchecked_mut<T: 'static>(&self) -> Option<Mut<'_, T>> {
+        let component_id = self.components.get_resource_id(TypeId::of::<T>())?;
+        self.get_non_send_unchecked_mut_with_id(component_id)
+    }
+
+    /// Temporarily removes the requested resource from this [World], then re-adds it before returning.
+    /// This enables safe mutable access to a resource while still providing mutable world access
+    /// ```
+    /// use bevy_ecs::world::World;
+    /// struct A(u32);
+    /// struct B(u32);
+    /// let mut world = World::new();
+    /// world.insert_resource(A(1));
+    /// let entity = world.spawn().insert(B(1)).id();
+    ///
+    /// world.resource_scope(|a: &mut A, world| {
+    ///     let b = world.get_mut::<B>(entity).unwrap();
+    ///     a.0 += b.0;
+    /// });
+    /// assert_eq!(world.get_resource::<A>().unwrap().0, 2);
+    /// ```
+    pub fn resource_scope<T: Component, U>(
+        &mut self,
+        f: impl FnOnce(&mut T, &mut World) -> U,
+    ) -> U {
+        let mut resource = self
+            .remove_resource::<T>()
+            .expect("resource does not exist");
+        let result = f(&mut resource, self);
+        self.insert_resource(resource);
+        result
+    }
```

Comment:
This flags the resource as added, we probably want to flag it as mutated instead.
Ideally it would track whether it was actually mutated, but that probably belongs to a future PR.

Reply:
Good catch. I think that is a logical error worth fixing now. We should probably read the flags, add `mutated`, then write them back.

Reply:
Done!